From be11ca24be64b87b424dc0660f5def2629b93cc1 Mon Sep 17 00:00:00 2001
From: Weiwei Wang <weiwei.wang@windriver.com>
Date: Mon, 29 Aug 2011 10:06:29 +0800
Subject: [PATCH] net/dpa: fix wrong free function for p_Fm->p_FmStateStruct

p_Fm->p_FmStateStruct is allocated by function XX_Malloc(), so XX_Free
needs to be used as free function, but not XX_FreeSmart() which corresponds
to XX_MallocSmart(). Otherwise, the call trace appears when do reboot:

Oops: Exception in kernel mode, sig: 5 [#1]
PREEMPT SMP NR_CPUS=8 LTT NESTING LEVEL : 0
P4080 DS
NIP: c0103b4c LR: c0498310 CTR: c00d6768
REGS: ea5f7ca0 TRAP: 0700   Not tainted  (2.6.34.10-grsec-WR4.3.0.0_cgl)
MSR: 00029002 <EE,ME,CE>  CR: 24004428  XER: 00000000
TASK = ea768b00[1048] 'kexec' THREAD: ea5f6000 CPU: 1
GPR00: 00000001 ea5f7d50 ea768b00 deadbeef c13c5b60 eb06f3c0 c0498310 00000000
GPR08: 00028000 00000000 00000040 00000000 f93de800 10037298 00000000 00000000
GPR16: 1002f2b8 1002f298 00000002 10030000 1002f2c0 00000002 10030000 00000000
GPR24: 10030000 00000002 bfa0cd94 00000000 ffffffff c0800000 deadbeef ebf91000
NIP [c0103b4c] kfree+0x118/0x18c
LR [c0498310] xx_Free+0x10/0x20
Call Trace:
[ea5f7d50] [c0497de0] XX_UnlockIntrSpinlock+0x10/0x20 (unreliable)
[ea5f7d70] [c0498310] xx_Free+0x10/0x20
[ea5f7d80] [c0498e2c] xx_FreeSmart+0x14/0x24
[ea5f7d90] [c0497f1c] XX_FreeSmart+0x10/0x20
[ea5f7da0] [c0429cd4] FM_Free+0x13c/0x1cc
[ea5f7db0] [c048edac] FreeFmDev+0x40/0xb8
[ea5f7dc0] [c048ee3c] fm_shutdown+0x18/0x2c
[ea5f7dd0] [c053aed0] of_platform_device_shutdown+0x30/0x40
[ea5f7de0] [c03cea50] device_shutdown+0x40/0xc8
[ea5f7df0] [c006318c] kernel_restart_prepare+0x30/0x44
[ea5f7e00] [c0088eac] kernel_kexec+0xd8/0xfc
[ea5f7e20] [c00633d4] sys_reboot+0x180/0x200
[ea5f7f40] [c0010f50] ret_from_syscall+0x0/0x4
  Exception: c00 at 0xff5ebe4
  LR = 0x1000301c
Instruction dump:
bba10014 7c0803a6 38210020 4e800020 80c10008 7fc5f378 4bffdca5 4bffffbc
80040000 54000422 3000ffff 54000ffe <0f000000> 7c832378 4bfd7599 4bffffa8

Signed-off-by: Weiwei Wang <weiwei.wang@windriver.com>
---
 drivers/net/dpa/NetCommSw/Peripherals/FM/fm.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/net/dpa/NetCommSw/Peripherals/FM/fm.c b/drivers/net/dpa/NetCommSw/Peripherals/FM/fm.c
index 4f69c6c..68e172a 100644
--- a/drivers/net/dpa/NetCommSw/Peripherals/FM/fm.c
+++ b/drivers/net/dpa/NetCommSw/Peripherals/FM/fm.c
@@ -3130,7 +3130,7 @@ t_Error FM_Free(t_Handle h_Fm)
     FreeInitResources(p_Fm);
 
     if(!p_Fm->recoveryMode)
-        XX_FreeSmart(p_Fm->p_FmStateStruct);
+        XX_Free(p_Fm->p_FmStateStruct);
 
     XX_Free(p_Fm);
 
-- 
1.7.0.4

